### PR TITLE
respect parsed tz instead of preferring tzlocal

### LINKED ISSUE #318: RFC822 timestamp using GMT is parsed into BST if local time is UK time
The following code snipped prints `True` unless the machine is using UK time.

```python
from datetime import datetime
from dateutil.parser import parse
from dateutil.tz import tzutc


d1 = datetime(2002, 10, 2, 13, 0, tzinfo=tzutc())
d2 = parse('Wed, 02 Oct 2002 13:00:00 GMT')
print(d1 == d2)
```

It appears that at [this point](https://github.com/dateutil/dateutil/blob/master/dateutil/parser.py#L605) in the parse function, the parser sees that local time is one of `('GMT', 'BST')` because in the Summer the UK uses BST (GMT+1).

Source issue: boto/botocore#1070

### PR DESCRIPTION
**Pull request by jstewmon**
if the computer is in a timezone where daylight savings is observed,
such as Europe/London, and a date string specifies a timezone, the
parsed datetime may be incorrect if the parsed datetime falls within
daylight savings but the date string specified the standard time

this is especially problematic for users in GMT/BST when parsing
dates in http or email headers, which are frequently specified as
GMT

fixes #318

### CODE REVIEW

File: dateutil/parser.py [hunk 1]
```diff
@@ -602,8 +602,6 @@ def parse(self, timestr, default=None, ignoretz=False, tzinfos=None, **kwargs):
                     raise ValueError("Offset must be tzinfo subclass, "
                                      "tz string, or int offset.")
                 ret = ret.replace(tzinfo=tzinfo)
-            elif res.tzname and res.tzname in time.tzname:
```
Comment (pganssle) [inline dateutil/parser.py]: This would break compatibility, I'm pretty sure we shouldn't do this as implemented. The most I'd be comfortable with is moving this down the list as the last resort rather than one of the first.

File: dateutil/test/test_parser.py [hunk 2]
```diff
@@ -125,11 +127,23 @@ def testDateCommandFormatWithLong(self):
                                    tzinfos={"BRST": long(-10800)}),
                              datetime(2003, 9, 25, 10, 36, 28,
                                       tzinfo=self.brsttz))
+
     def testDateCommandFormatIgnoreTz(self):
         self.assertEqual(parse("Thu Sep 25 10:36:28 BRST 2003",
                                ignoretz=True),
                          datetime(2003, 9, 25, 10, 36, 28))
 
+    def testDateCommandFormatRespectTz(self):
```
Comment (pganssle) [inline dateutil/test/test_parser.py]: Pretty sure this will always fail on Windows, so this would need to be changed in the following ways:

1. Add a `@skipIf` to skip this test on Windows.
2. Use [`dateutil.tests._common.TZEnvContext`](https://github.com/dateutil/dateutil/blob/master/dateutil/test/_common.py#L159) for this, rather than just setting the `os.environ` directly like this.

File: dateutil/parser.py (hunk 1 above)
Comment (jstewmon) [inline dateutil/parser.py]: I'm not sure how this case would be hit if it's moved to the bottom.

Is the compatibility case that `res.tzname` is not `None`, but res.tzoffset is `None`?

Can you provide a date string that I can use to add a test for the compatibility concern?

File: dateutil/test/test_parser.py (hunk 2 above)
Comment (jstewmon) [inline dateutil/test/test_parser.py]: Ah, thanks for the pointer here. I made the test choose the right context manager and timezone name by platform.

File: dateutil/parser.py (hunk 1 above)
Comment (pganssle) [inline dateutil/parser.py]: I think I meant more in the abstract, I didn't really think much about the implementation. Really, though, I think a better fix is going to be somewhat more complicated, as I mentioned in #318. Basically, I think what should be done is something like this (eliding over the details because I haven't though through the logic fully):

```python
# Code to determine if it's supposed to be `tzlocal()`
if isinstance(ret.tzinfo, tz.tzlocal):
    if ret.tzname() != res.tzname or not tz.datetime_exists(ret):
        if tz.datetime_ambiguous(ret) and tz.enfold(ret, fold=1).tzname() == res.tzname:
            ret = tz.enfold(ret, fold=1)
        elif res.tzoffset == 0:
            ret = ret.replace(tzinfo=tz.tzutc())
        elif res.tzoffset:
            ret = ret.replace(tzinfo=tz.tzoffset(res.tzname, res.tzoffset))
```

It probably needs to be a bit more complicated than that to handle all the edge cases properly, and to disambiguate ambiguous dates, but the logic is at least somewhat clear:

1. If it could be parsed as `tzlocal()` but applying `tzlocal()` to the date doesn't give you the input `tzname` you expected (likely because of a DST mismatch), check if it's just an ambiguous date.
2. If it's an ambiguous date, keep `tzlocal()` and set the `fold` attribute to 1 to disambiguate per PEP-495.
3. If it's not an ambiguous date, try all the other possible ways to parse the date that aren't `tzlocal()`, and if you find a good one, apply it.
4. If you didn't find a good one, that's an unpleasant situation, but to maintain backwards compatibility, just leave `tzlocal()` attached (we can add a `FutureWarning` if need be on this).

I'm thinking the best implementation would be to instead of using `if / elif`, after the `tzinfos` part has been parsed, just start a queue of possible time zones, so like in this case the queue would be something like `[tzlocal(), `tzutc()`], then assign the first one in the list that creates a non-imaginary date, and if none of them create non-imaginary dates, return the first one in the list.

Comment (jstewmon) [inline dateutil/parser.py]: Ok, thanks for the feedback. You have more context than me on the issue, so I'll defer to your expertise, but I thought about the priority a little differently:

1. If an offset was parsed, it should always be the exact offset in the `tzinfo`.
  - If the offset is 0, `tzinfo=tzutc()`, else `tzinfo=tz.tzoffset(res.tzname, res.tzoffset)`
2. Otherwise, if an name was parsed and the name is in `time.tzname`, use the offset for the explicitly named zone.

```python
            elif res.tzoffset == 0:
                ret = ret.replace(tzinfo=tz.tzutc())
            elif res.tzoffset:
                ret = ret.replace(tzinfo=tz.tzoffset(res.tzname, res.tzoffset))
            elif res.tzname:
                try:
                    tzindex = time.tzname.index(res.tzname)
                except ValueError:
                    pass
                else:
                    offset = tzindex and time.altzone or time.timezone
                    ret = ret.replace(tzinfo=tz.tzoffset(res.tzname, -offset))
                    ret = ret.astimezone(tz.tzlocal())
```

This assumes the date is not ambiguous if the date string has a timezone.

If you don't think this is right, I'm happy to let you take over and resolve however you think is best.

Comment (jstewmon) [inline dateutil/parser.py]: After thinking more about the problem, I agree with you.

I was originally thinking about the problem for scenarios where a date string is produced by a machine - in such cases, we can probably expect that a timezone in a date string is significant and should be honored. For example, the date string `'2016-10-10T08:00:00 CST'` means `2016-10-10 08:00:00-06:00` because `CST` is `UTC-06:00`.

However, in many cases the source of a date string may be from user input where the user considers the time to be authoritative and the timezone to be informative but not necessarily correct. I would call this a "colloquial" date string. For example, a user may write `October 10, 8am CST`, expecting the time to be `2016-10-10T08:00:00-05:00` because CST was given to mean US Central time, not exactly Central Standard Time.

My previous proposal would violate the expectations for colloquial date strings because `CST` would be taken explicitly.

I'm now wondering if the behavior of dateutil actually needs to change for this. Similarly to your suggestion on #318, the `parse` caller could generally enforce the offset of local timezones by passing `tzinfos={time.tzname[0]: -time.timezone, time.tzname[1]: -time.altzone}`. Having these examples in the documentation would probably be sufficient to avoid issues like the one that was observed by botocore users in the UK.

Comment (pganssle) [inline dateutil/parser.py]: I think it's reasonable for the behavior to change just so that by default the closest to the "right" thing is inferred from whatever context we have, but I agree with you and would be very happy for any PRs that improve the documentation. If you don't have time, I will make sure that your suggested documentation goes somewhere in there - I think the docs are in need of a major re-organization and cleanup anyway, as they are somewhat hard to follow.

I'm going to leave this PR open for now because I'd like to branch off of it to implement the functionality I described above, and I can use a modified version of the test you added.

**Comment by pganssle** [comment]
New PR at #517
